### ISSUE TITLE
Misleading error message in case of ref-returning this in a reference type

### ISSUE THREAD

**Issue by alrz** [dotnet/roslyn#15704]
**Version Used**:  https://github.com/dotnet/roslyn/commit/0694d5ccc05b7e1cc37c82a1c15c8f9fc23e3775

**Steps to Reproduce**:

```cs
public class C 
{
    public ref C M()
    {
        return ref this;
    }
}
```
**Expected Behavior**:

`error CS1605: Cannot use 'this' as a ref or out value because it is read-only`

**Actual Behavior**:

`error CS8170: Struct members cannot return 'this' or other instance members by reference`

**Comment by jcouv** [comment]
FYI @VSadov

**Comment by gafter** [comment]
Actually, I think the existing error is not bad. If we add support for readonly refs, this would still be an error for the reason mentioned in the diagnostic.

**Comment by alrz** [comment]
I don't get it. the error is clearly misleading as I'm not using a struct. What you say applies to "expected" message, not actual.

**Comment by gafter** [comment]
Yes, I see you're right.
